packages/@atjson/hir/src/hir.ts

```diff
@@ -16,17 +21,64 @@ function compareAnnotations(a: Annotation, b: Annotation) {
 
 export default class HIR {
   rootNode: HIRNode;
+  sliceNodes: Record<string, HIRNode>;
 
   constructor(doc: Document) {
     let document: Document = doc.clone();
 
+    // Do some basic document normalization here,
+    // expanding any zero-width annotations
     for (let a of document.annotations) {
       if (a.start === a.end) {
         document.insertText(a.start, "\uFFFC");
         a.start = Math.max(0, a.start - 1);
       }
     }
 
+    let slices: Record<string, Document> = {};
+    let sliceRanges: { start: number; end: number }[] = [];
+    let sliceAnnotations = document.where((a) => is(a, SliceAnnotation));
+    for (let annotation of sliceAnnotations) {
+      let slice = document.clone();
+      slice
+        .where(
+          (a) =>
+            !(a.start >= annotation.start && a.end <= annotation.end) ||
+            a.id === annotation.id
+        )
+        .remove();
+      slices[annotation.id] = slice.slice(annotation.start, annotation.end);
```

Comment:
`document.slice` returns a new document, so I think you can first slice and then remove the annotations from the new document to avoid the extra call to `document.clone`

Reply:
I fixed this.